shared/typeinference/codeql/typeinference/internal/TypeInference.qll

```diff
@@ -0,0 +1,776 @@
+        exists(TypeMention immediateBaseMention |
+          pragma[only_bind_into](immediateBaseMention) =
+            getABaseTypeMention(pragma[only_bind_into](sub))
+        |
+          // immediate base class
+          baseMention = immediateBaseMention and
+          t = immediateBaseMention.resolveTypeAt(path)
+          or
+          // transitive base class
+          exists(Type immediateBase | immediateBase = resolveTypeMentionRoot(immediateBaseMention) |
+            baseTypeMentionHasNonTypeParameterAt(immediateBase, baseMention, path, t)
+            or
+            exists(TypePath path0, TypePath prefix, TypePath suffix, TypeParameter tp |
+              /*
+               * Example:
+               *
+               * - `prefix = "0.0"`,
+               * - `path0 = "0"`,
+               * - `suffix = ""`,
+               * - `path = "0.0"`
+               *
+               * ```csharp
+               * class C<T1> { }
+               *       ^ `t`
+               *
+               * class Base<T2> { }
+               *
+               * class Mid<T3> : Base<C<T3>> { }
+               * //    ^^^ `immediateBase`
+               * //        ^^ `tp`
+               * //              ^^^^^^^^^^^ `baseMention`
+               *
+               * class Sub<T4> : Mid<C<T4>> { }
+               * //    ^^^ `sub`
+               * //              ^^^^^^^^^^ `immediateBaseMention`
+               * ```
+               */
+
+              baseTypeMentionHasTypeParameterAt(immediateBase, baseMention, prefix, tp) and
+              t = immediateBaseMention.resolveTypeAt(path0) and
+              path0.isCons(tp, suffix) and
+              path = prefix.append(suffix)
+            )
+          )
+        )
+      }
+    }
+
+    private import BaseTypes
+
+    /** Provides the input to `Matching`. */
+    signature module MatchingInputSig {
+      /**
+       * A position inside a declaration. For example, the integer position of a
+       * parameter inside a method or the return type of a method.
+       */
+      bindingset[this]
+      class DeclarationPosition {
+        /** Gets a textual representation of this position. */
+        bindingset[this]
+        string toString();
+      }
+
+      /** A declaration, for example a method. */
+      class Declaration {
+        /** Gets a textual representation of this declaration. */
+        string toString();
+
+        /** Gets the location of this declaration. */
+        Location getLocation();
+
+        /** Gets the type parameter at position `tppos` of this declaration, if any. */
+        TypeParameter getTypeParameter(TypeParameterPosition tppos);
+
+        /**
+         * Gets the declared type of this declaration at `path` for position `dpos`.
+         *
+         * For example, if this declaration is the method `int M(bool b)`,
+         * then the declared type at parameter position `0` is `bool` and the
+         * declared return type is `int`.
```

Comment:
Could we elaborate the example concerning the following? If `M` is an instance method then presumably there's also a `DeclarationPosition` for the instance argument? If not, then the `new Sub<int>().ToString();` example below in `hasBaseTypeMention` would not qualify for `predicate relevantAccess`.